File: app/src/main/java/com/aravi/dot/activities/log/LogsActivity.java
/*
 * Copyright (C) 2020.  Aravind Chowdary (@kamaravichow)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.aravi.dot.activities.log;

import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;

import com.aravi.dot.R;
import com.aravi.dot.adapter.LogAdapter;
import com.aravi.dot.databinding.ActivityLogsBinding;
import com.aravi.dot.model.Logs;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;

public class LogsActivity extends AppCompatActivity {
    private ActivityLogsBinding mBinding;
    private LogsViewModel mLogsViewModel;
    private List<Logs> logsList;
    private LogAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivityLogsBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
        mLogsViewModel = ViewModelProviders.of(this).get(LogsViewModel.class);
        setSupportActionBar(mBinding.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        init();
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return super.onSupportNavigateUp();
    }


    @Override
    protected void onStart() {
        // Hides the clear button and shows progress bar on start
        mBinding.clearLogsButton.hide();
        mBinding.progressBar.setVisibility(View.VISIBLE);
        super.onStart();
    }

    private void init() {
        // Creates a new array-list object on initialisation
        logsList = new ArrayList<>();

        // I've set the linear layout manager in the layout file
        // mBinding.logsRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        // get all the log list live data from the view model
        mLogsViewModel.getmLogsList().observe(this, logs -> {
            mBinding.progressBar.setVisibility(View.INVISIBLE);
            adapter = new LogAdapter(LogsActivity.this, logs);
            mBinding.logsRecyclerView.setAdapter(adapter);
            if (logs.isEmpty()) {
                mBinding.clearLogsButton.hide();
                findViewById(R.id.emptyListImage).setVisibility(View.VISIBLE);
            } else {
                mBinding.clearLogsButton.show();
                mBinding.clearLogsButton.setOnClickListener(v -> {
                    mLogsViewModel.clearLogs();
                    showSnackBar("Logs Cleared");
                });
                mBinding.emptyListImage.setVisibility(View.INVISIBLE);
            }
        });
    }

    /**
     * Snack-bar Method
     * Gets String As Input
     */
    private void showSnackBar(String message) {
        Snackbar.make(mBinding.getRoot(), message, Snackbar.LENGTH_SHORT).show();
    }


}

File: app/src/main/java/com/aravi/dot/activities/log/LogsActivity.kt
/*
 * Copyright (C) 2020.  Aravind Chowdary (@kamaravichow)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.aravi.dot.activities.log

import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.aravi.dot.Constants
import com.aravi.dot.R
import com.aravi.dot.activities.log.adapter.LogsAdapter
import com.aravi.dot.databinding.ActivityLogsBinding
import com.aravi.dot.util.PermissionUtils
import com.aravi.dot.util.Utils
import me.aravi.commons.base.BaseActivity
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel


class LogsActivity : BaseActivity() {
    private lateinit var binding: ActivityLogsBinding
    private val viewModel: LogsViewModel by viewModel()
    private lateinit var adapter: LogsAdapter

    private val utils: Utils by inject()
    private val permissionUtils: PermissionUtils by inject()


    var permission: String = Constants.PERMISSION_CAMERA

    private lateinit var bundle: Bundle


    override fun onStart() {
        // Hides the clear button and shows progress bar on start
        binding.clearLogsButton.hide()
        super.onStart()
    }


    override val contentView: View
        get() {
            binding = ActivityLogsBinding.inflate(layoutInflater)
            return (binding.root)
        }

    override fun onViewReady(savedInstanceState: Bundle?, intent: Intent?) {
        transparentStatusBar()
        binding.appbar.setPadding(0, statusBarHeight(resources) + 10, 0, 0)

        setSupportActionBar(binding.toolbar)
        bundle = ActivityOptions
            .makeCustomAnimation(this, R.anim.slide_out_left, R.anim.slide_in_right)
            .toBundle();
        setBack()

        init()

    }

    private fun init() {
        val bundle = intent.extras
        if (bundle != null) {
            if (bundle.containsKey(Constants.EXTRA_PERMISSION)) {
                permission = bundle.getString(Constants.EXTRA_PERMISSION)!!
            }
        }



        supportActionBar!!.title = ""
        binding.logsRecyclerView.layoutManager = LinearLayoutManager(this)
        adapter = LogsAdapter(this, permission, utils, permissionUtils)
        binding.logsRecyclerView.adapter = adapter

        binding.logsRecyclerView.isRefreshing = true
        viewModel.loadList(permission).observe(this) {
            adapter.stopLoading();
            adapter.setLogsList(it);

            if (it.isEmpty()) {
                binding.emptyListImage.visibility = View.VISIBLE

            } else {
                binding.emptyListImage.visibility = View.GONE
                binding.logsRecyclerView.addFooterView(R.layout.item_log_footer)
            }
            binding.logsRecyclerView.removeAllFooterView()
            binding.logsRecyclerView.isRefreshing = false
        }

//        binding.logsRecyclerView.setOnRefreshListener {
//            adapter.stopLoading();
//            init()
//        }

    }


    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }

    override fun onBackPressed() {
        super.onBackPressed()
    }

//    override fun finish() {
//        super.finish()
//        overridePendingTransition(R.anim.slide_out_right, R.anim.slide_in_right)
//    }
}
